pkg/controller/master/kubevirt/vm_migration_network.go

```
@@ -0,0 +1,252 @@
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.VMMigrationNetworkSettingName {
+		return setting, nil
+	}
+	settingCopy := setting.DeepCopy()
+
+	if settingCopy.Annotations == nil {
+		if settingCopy.Value == "" {
+			// initialization case, don't update status, just skip it.
+			return setting, nil
+		}
+		settingCopy.Annotations = make(map[string]string)
+	}
+
+	if h.checkIsSameHashValue(setting) {
+		return setting, nil
+	}
+
+	logrus.Infof("vm migration network change: %s", settingCopy.Value)
+
+	kubevirt, err := h.kubevirtCache.Get(util.HarvesterSystemNamespaceName, util.KubeVirtObjectName)
+	if err != nil {
+		return setting, fmt.Errorf("get kubevirt error %v", err)
+	}
+	kubevirtCopy := kubevirt.DeepCopy()
+	if kubevirtCopy.Spec.Configuration.MigrationConfiguration == nil {
+		kubevirtCopy.Spec.Configuration.MigrationConfiguration = &kubevirtv1.MigrationConfiguration{}
+	}
+
+	oldNadAnnotation := setting.Annotations[VMMigrationNetworkNadAnnotation]
+	removeOldNad := true
+
+	var nadAnnotation string
+	if setting.Value != "" {
+		nad, err := h.findOrCreateNad(setting)
+		if err != nil {
+			return setting, err
+		}
+		nadAnnotation = fmt.Sprintf("%s/%s", nad.Namespace, nad.Name)
+		if oldNadAnnotation == nadAnnotation {
+			removeOldNad = false
+		}
+
+		kubevirtCopy.Spec.Configuration.MigrationConfiguration.Network = &nad.Name
+	} else {
+		kubevirtCopy.Spec.Configuration.MigrationConfiguration.Network = nil
+	}
+
+	if !reflect.DeepEqual(kubevirt.Spec.Configuration, kubevirtCopy.Spec.Configuration) {
+		if _, err = h.kubevirts.Update(kubevirtCopy); err != nil {
+			return setting, fmt.Errorf("update kubevirt error %v", err)
+		}
+	}
+
+	if removeOldNad {
+		oldNadNamespaceAndName := strings.Split(oldNadAnnotation, "/")
+		if len(oldNadNamespaceAndName) == 2 {
+			h.removeOldNad(oldNadNamespaceAndName[0], oldNadNamespaceAndName[1])
```

we do not return an error and fail with a warning, should we return an error or just add more retry logic in the `removeOldNad` function before we log the error?

---

Return error if the handler cannot remove old NAD. Thanks.